server/src/main/java/org/elasticsearch/cluster/ClusterInfo.java

```diff
@@ -133,8 +156,14 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         }
         builder.endObject(); // end "nodes"
         builder.startObject("shard_sizes"); {
-            for (ObjectObjectCursor<String, Long> c : this.shardSizes) {
-                builder.humanReadableField(c.key + "_bytes", c.key, new ByteSizeValue(c.value));
+            for (ObjectObjectCursor<ShardId, Long> c : this.shardSizes) {
+                final String shardKey;
+                if (BWC_SHARD_ID_UUID.equals(c.key.getIndex().getUUID())) {
+                    shardKey = c.key.getIndexName();
+                } else {
+                    shardKey = c.key.toString();
+                }
+                builder.humanReadableField(shardKey + "_bytes", shardKey, new ByteSizeValue(c.value));
```

Comment:
This is arguably a breaking change since we will no longer include the `[p]` or `[r]` suffix in the shard sizes reported by the allocation explain output. We could fake these out for BWC, reporting the size of each shard twice, and allowing users to opt-in to the future behaviour here with a system property. TBD.

Reply:
The allocation explain API is mostly meant for human consumption. As long as the field names are preserved in the response output, things should be ok.

Reply:
As it stands, this PR changes the field names in the response output, because it drops the `[p]` or `[r]` marker from `shardKey`. Even if we were to track all copies then I think we'd also need to change these field names.

Reply:
I see. I think breaking this is fine.